modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/rule/logical/IgniteMultiJoinOptimizeBushyRule.java

```diff
@@ -0,0 +1,449 @@
+ * from best plans, until all relations are joined.
+ *
+ * <p>Current limitations are as follow:<ol>
+ *     <li>Only INNER joins are supported</li>
+ *     <li>Number of relations to optimize is limited to 20. This is due to time and memory complexity of algorithm chosen.</li>
+ *     <li>Disjunctive predicate is not considered as connections.</li>
+ * </ol>
+ */
+@Value.Enclosing
+public class IgniteMultiJoinOptimizeBushyRule
+        extends RelRule<IgniteMultiJoinOptimizeBushyRule.Config>
+        implements TransformationRule {
+
+    private static final int MAX_JOIN_SIZE = 20;
+
+    /**
+     * Comparator that puts better vertexes first.
+     *
+     * <p>Better vertex is the one that incorporate more relations, or costs less.
+     */
+    private static final Comparator<Vertex> VERTEX_COMPARATOR = 
+            Comparator.<Vertex>comparingInt(v -> bitCount(v.id))
+                    .reversed()
+                    .thenComparingDouble(v -> v.cost);
+
+    /** Creates a MultiJoinOptimizeBushyRule. */
+    private IgniteMultiJoinOptimizeBushyRule(Config config) {
+        super(config);
+    }
+
+    @Override
+    public void onMatch(RelOptRuleCall call) {
+        MultiJoin multiJoinRel = call.rel(0);
+
+        int numberOfRelations = multiJoinRel.getInputs().size();
+        if (numberOfRelations > MAX_JOIN_SIZE) {
+            return;
+        }
+
+        // Currently, algorithm below can handle only INNER JOINs
+        if (multiJoinRel.isFullOuterJoin()) {
+            return;
+        }
+
+        for (JoinRelType joinType : multiJoinRel.getJoinTypes()) {
+            if (joinType != JoinRelType.INNER) {
+                return;
+            }
+        }
+
+        LoptMultiJoin multiJoin = new LoptMultiJoin(multiJoinRel);
+
+        RexBuilder rexBuilder = multiJoinRel.getCluster().getRexBuilder();
+        RelBuilder relBuilder = call.builder();
+        RelMetadataQuery mq = call.getMetadataQuery();
+
+        List<RexNode> unusedConditions = new ArrayList<>();
+
+        Int2ObjectMap<List<Edge>> edges = collectEdges(multiJoin, unusedConditions);
+        Int2ObjectMap<Vertex> bestPlan = new Int2ObjectOpenHashMap<>();
+        BitSet connections = new BitSet(1 << numberOfRelations);
+
+        int id = 0b1;
+        int fieldOffset = 0;
+        for (RelNode input : multiJoinRel.getInputs()) {
+            TargetMapping mapping = Mappings.offsetSource(
+                    Mappings.createIdentity(input.getRowType().getFieldCount()),
+                    fieldOffset,
+                    multiJoin.getNumTotalFields()
+            );
+
+            bestPlan.put(id, new Vertex(id, mq.getRowCount(input), input, mapping));
+            connections.set(id);
+
+            id <<= 1;
+            fieldOffset += input.getRowType().getFieldCount();
+        }
+
+        Vertex bestSoFar = null;
+        for (int k = 2; k <= numberOfRelations; k++) {
+            for (int s = (1 << (k - 1)) + 1; s < 1 << k; s++) {
```

Comment:
```suggestion
        // Nested for-loops generates sequence from 0b11 up to  (1<< numberOfRelations) - 1 skipping values, which are power of 2.
        for (int k = 2; k <= numberOfRelations; k++) {
            for (int s = (1 << (k - 1)) + 1; s < 1 << k; s++) {
```

Reply:
Simplified this part a bit: replaced these two for-loops with single loop with condition